pkg/kbagent/service/action_utils.go

```diff
@@ -0,0 +1,614 @@
+	}()
+
+	cmd := exec.CommandContext(ctx, action.Commands[0], mergedArgs...)
+	if len(mergedEnv) > 0 {
+		cmd.Env = mergedEnv
+	}
+
+	cmd.Stdin = stdinReader
+	cmd.Stdout = stdoutWriter
+	cmd.Stderr = stderrWriter
+
+	go func() {
+		defer cancel()
+		defer close(errChan)
+
+		if err := cmd.Start(); err != nil {
+			if errors.Is(ctx.Err(), context.DeadlineExceeded) {
+				errChan <- kbaproto.ErrTimedOut
+			} else {
+				errChan <- errors.Wrapf(kbaproto.ErrFailed, "failed to start command: %v", err)
+			}
+			return
+		}
+
+		execErr := cmd.Wait()
+		if execErr != nil {
+			if errors.Is(ctx.Err(), context.DeadlineExceeded) {
+				execErr = kbaproto.ErrTimedOut
+			}
+		}
+		errChan <- execErr
+	}()
+
+	return nil
+}
+
+func httpActionCallX(ctx context.Context, cancel context.CancelFunc,
+	action *kbaproto.HTTPAction, parameters map[string]string, errChan chan error, _ io.Reader, stdoutWriter, stderrWriter io.Writer) error {
+	var (
+		// TODO: resolve the port
+		cli         = httpClient(action.Host, action.Port)
+		method, url = httpActionMethodNURL(action)
+	)
+	body, err1 := renderTemplateData("http body", parameters, action.Body)
+	if err1 != nil {
+		return err1
+	}
+	req, err2 := http.NewRequestWithContext(ctx, method, url, strings.NewReader(body))
+	if err2 != nil {
+		return err2
+	}
+
+	for _, h := range action.Headers {
+		val, err3 := renderTemplateData("http header", parameters, h.Value)
+		if err3 != nil {
+			return err3
+		}
+		req.Header.Add(h.Name, val)
+	}
+
+	go func() {
+		defer cancel()
+		defer close(errChan)
+
+		handleError := func(err error, msg string) {
+			if errors.Is(ctx.Err(), context.DeadlineExceeded) {
+				errChan <- kbaproto.ErrTimedOut
+			} else {
+				errChan <- errors.Wrapf(kbaproto.ErrFailed, "%s: %v", msg, err)
+			}
+		}
+
+		rsp, err4 := cli.Do(req)
+		if err4 != nil {
+			handleError(err4, "failed to issue http request")
+			return
+		}
+		defer safeClose(rsp.Body)
+
+		output, err5 := io.ReadAll(rsp.Body)
+		if err5 != nil {
+			handleError(err5, "failed to read http response")
+			return
+		}
+
+		succeed := rsp.StatusCode >= 200 && rsp.StatusCode < 300
+		if succeed {
+			if len(output) > 0 && stdoutWriter != nil {
+				_, err6 := stdoutWriter.Write(output)
+				if err6 != nil {
+					handleError(err6, "failed to write http response")
+					return
+				}
+			}
+			errChan <- nil
+		} else {
+			if len(output) > 0 && stderrWriter != nil {
+				_, err7 := stderrWriter.Write(output)
+				if err7 != nil {
+					handleError(err7, "failed to write http response")
+					return
+				}
+			}
+			errChan <- errors.Wrapf(kbaproto.ErrFailed, "http request failed, status: %s", rsp.Status)
+		}
+	}()
+
+	return nil
+}
+
+func httpClient(host, port string) *http.Client {
+	key := fmt.Sprintf("http://%s:%s", host, port)
+	if v, ok := clientCache.Load(key); ok {
+		return v.(*http.Client)
+	}
+	cli, _ := clientCache.LoadOrStore(key, &http.Client{
+		Transport: defaultHTTPTransport,
+	})
+	return cli.(*http.Client)
```

Comment:
Can a http.Client be used by many requests? In that case, a global single Client would be enough.